Strange behaviour in sysrepo when mandatory=true leafs of different models

### Things you may try first
(put "x" in "[]" if you already tried following)
* [] Did you check if this is a duplicate issue?
* [] Did you test it on the latest sysrepo devel branch?

### Description
Two YANG models are in use: ietf-interfaces (plus all augments of ietf-ip and others) plus another YANG model defined by me (routing-indra.yang) that has augments for several ietf-models.
Using sysrepocfg I try to merge an XML file where I configure a new IPv6 address in the ietf-interfaces model and the sysrepo complains that the merge cannot be done because there is a missing field. This field has the mandatory as TRUE but it is located in the model defined by me that has nothing to do with the ietf-models.
Moreover, when I perform a merge with anothef XML file that only affects the routing-indra model, then the sysrepod report successful merging (as expected)
I don't know if it is a bug of the sysrepo that I am using a this moment or not.

### Steps to Reproduce
1. [First Step]
Install the ietf-interfaces, ietf-ip and routing-indra yang models (see the output of the sysrepoctl --list)
cuda1@cuda1:~> sysrepoctl --list
Sysrepo schema directory: /etc/sysrepo/yang/
Sysrepo data directory:   /etc/sysrepo/data/
(Do not alter contents of these directories manually)

Module Name                | Revision   | Conformance | Data Owner          | Permissions | Submodules                    | Enabled Features
-----------------------------------------------------------------------------------------------------------------------------------------------
ietf-netconf-notifications | 2012-02-06 | Installed   | root:root           | 666         |                               |
ietf-netconf               | 2011-06-01 | Installed   | root:root           | 666         |                               | writable-running candidate rollback-on-error validate startup xpath url
ietf-netconf-acm           | 2012-02-22 | Imported    |                     |             |                               |
nc-notifications           | 2008-07-14 | Installed   | root:root           | 666         |                               |
notifications              | 2008-07-14 | Installed   | root:root           | 666         |                               |
iana-if-type               | 2014-05-08 | Installed   |                     |             |                               |
ietf-interfaces            | 2014-05-08 | Installed   | root:root           | 666         |                               | if-mib
ietf-ip                    | 2014-06-16 | Installed   |                     |             |                               |
ietf-netconf-with-defaults | 2011-06-01 | Installed   |                     |             |                               |
ietf-netconf-monitoring    | 2010-10-04 | Installed   | root:root           | 666         |                               |
ietf-yang-library          | 2018-01-17 | Installed   | root:root           | 666         |                               |
ietf-yang-types            | 2013-07-15 | Installed   |                     |             |                               |
turing-machine             | 2013-12-27 | Installed   | root:root           | 666         |                               |
ietf-datastores            | 2017-08-17 | Installed   |                     |             |                               |
ietf-routing               | 2015-10-16 | Installed   | root:root           | 666         |                               |
ietf-ipv4-unicast-routing  | 2015-10-16 | Installed   |                     |             |                               |
ietf-ipv6-unicast-routing  | 2015-10-16 | Installed   |                     |             |                               |
ietf-qos-policy            | 2016-03-03 | Installed   | root:root           | 666         |                               |
ietf-qos-classifier        | 2016-03-03 | Installed   | root:root           | 666         |                               |
ietf-qos-action            | 2016-06-15 | Installed   | root:root           | 666         |                               |
ietf-access-control-list   | 2016-10-12 | Installed   | root:root           | 666         |                               |
ietf-packet-fields         | 2016-10-12 | Installed   |                     |             |                               |
ietf-inet-types            | 2013-07-15 | Installed   |                     |             |                               |
ietf-event-notifications   | 2016-10-27 | Installed   | root:root           | 666         |                               |
ydk                        | 2016-02-26 | Installed   | root:root           | 666         |                               |
routing-indra             | 2018-09-18 | Installed   | root:root           | 666         |                               | *

2. Run the command
cuda1@cuda1:~> sudo sysrepocfg -d running -l 4 --format=xml --merge=ipv6_merge.xml ietf-interfaces
[DBG] Setting log level for stderr logs to 4.
[DBG] Connecting to Sysrepo Engine.
[DBG] Connecting to socket=/var/run/sysrepod.sock
[INF] Connected to daemon Sysrepo Engine at socket=/var/run/sysrepod.sock
[DBG] Sending version-verify request.
[DBG] version-verify request sent, waiting for response.
[DBG] version-verify response received, processing.
[DBG] Sending session-start request.
[DBG] session-start request sent, waiting for response.
[DBG] session-start response received, processing.
[DBG] Sending check-enabled-running request.
[DBG] check-enabled-running request sent, waiting for response.
[DBG] check-enabled-running response received, processing.
[DBG] Loading module schema: '/etc/sysrepo/yang/ietf-interfaces@2014-05-08.yang'.
[DBG] Sending discard-changes request.
[DBG] discard-changes request sent, waiting for response.
[DBG] discard-changes response received, processing.
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] No items found for xpath '/ietf-routing:*//.'
[WRN] libyang: Missing required element "max-ip-routes" in "license-params". (/routing-indra:routing-service-config/license-params)
[ERR] Input data are not valid: Missing required element "max-ip-routes" in "license-params". (/routing-indra:routing-service-config/license-params)
Errors were encountered during importing. Cancelling the operation.
[DBG] Sending session-stop request.
[DBG] session-stop request sent, waiting for response.
[DBG] session-stop response received, processing.




3.Run the command
cuda1@cuda1:~> sudo sysrepocfg -d running -l 4 --format=xml --merge=rt_srv.xml routing-indra
[sudo] password for root: 
[DBG] Setting log level for stderr logs to 4.
[DBG] Connecting to Sysrepo Engine.
[DBG] Connecting to socket=/var/run/sysrepod.sock
[INF] Connected to daemon Sysrepo Engine at socket=/var/run/sysrepod.sock
[DBG] Sending version-verify request.
[DBG] version-verify request sent, waiting for response.
[DBG] version-verify response received, processing.
[DBG] Sending session-start request.
[DBG] session-start request sent, waiting for response.
[DBG] session-start response received, processing.
[DBG] Sending check-enabled-running request.
[DBG] check-enabled-running request sent, waiting for response.
[DBG] check-enabled-running response received, processing.
[DBG] Loading module schema: '/etc/sysrepo/yang/routing-indra@2018-09-18.yang'.
[DBG] Sending discard-changes request.
[DBG] discard-changes request sent, waiting for response.
[DBG] discard-changes response received, processing.
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] No items found for xpath '/ietf-routing:*//.'
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] All items has been read for xpath '/ietf-interfaces:*//.'
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] Sending get-items request.
[DBG] get-items request sent, waiting for response.
[DBG] get-items response received, processing.
[DBG] All items has been read for xpath '/routing-indra:*//.'
[DBG] No changes were made.
The new configuration was successfully applied.
[DBG] Sending session-stop request.
[DBG] session-stop request sent, waiting for response.
[DBG] session-stop response received, processing.



**Expected behavior:**
 when merging the ipv6_merge.xml the sysrepo debug logs should not display:
[WRN] libyang: Missing required element "max-ip-routes" in "license-params". (/routing-indra:routing-service-config/license-params)
[ERR] Input data are not valid: Missing required element "max-ip-routes" in "license-params". (/routing-indra:routing-service-config/license-params)
Errors were encountered during importing. Cancelling the operation.

**Actual behavior:**
The ipv6 address cannot be configured because of this strange problem.

### Versions
Please provide versions for software needed to reproduce the bug:
Sysrepo version: 0.7.7
libyang version: 0.16.108
netopeer2-server: 0.7.3

### Attachments

[ipv6_merge.xml.txt](https://github.com/sysrepo/sysrepo/files/3077758/ipv6_merge.xml.txt)
[logs commands.txt](https://github.com/sysrepo/sysrepo/files/3077759/logs.commands.txt)
[routing-indra@2018-09-18_yang.txt](https://github.com/sysrepo/sysrepo/files/3077760/routing-indra%402018-09-18_yang.txt)
[rt_svr.xml.txt](https://github.com/sysrepo/sysrepo/files/3077761/rt_svr.xml.txt)

After having done a pull of libyang, sysrepo and netopeer 2-server now I have problems in compilation.
Libyang is successfully compiled (make) and installed (make install). However, when doing the same with master branch of sysrepo this is the errors displayed with doing 'make':
                           ^
/home/cuda1/git/sysrepo/src/common/sr_utils.c:705:27: note: each undeclared identifier is reported only once for each function it appears in
/home/cuda1/git/sysrepo/src/common/sr_utils.c: In function ?sr_libyang_anydata_copy_value?:
/home/cuda1/git/sysrepo/src/common/sr_utils.c:1324:10: error: ?LYD_ANYDATA_LYBD? undeclared (first use in this function)
     case LYD_ANYDATA_LYBD:
          ^
/home/cuda1/git/sysrepo/src/common/sr_utils.c:1325:10: error: ?LYD_ANYDATA_LYB? undeclared (first use in this function)
     case LYD_ANYDATA_LYB:
          ^
In file included from /home/cuda1/git/sysrepo/src/common/sr_common.h:47:0,
                 from /home/cuda1/git/sysrepo/src/common/sr_utils.c:40:
/home/cuda1/git/sysrepo/src/common/sr_utils.c: In function ?sr_subtree_to_dt?:
/home/cuda1/git/sysrepo/src/common/sr_utils.c:1933:92: error: too many arguments to function ?ly_errmsg?

And now I am stuck here.... Any idea of what is happening?

Thanks again

--- a/src/data_manager.c
+++ b/src/data_manager.c
@@ -2844,6 +2844,7 @@ dm_validate_data_info(dm_ctx_t *dm_ctx, dm_session_t *session, dm_data_info_t *i
     sr_list_t *data_for_validation = NULL;
     dm_tmp_ly_ctx_t *tmp_ctx = NULL;
     dm_data_info_t *dep_di = NULL;
+    const struct lys_module *mod;
     struct lyd_node *data_tree = NULL;
     bool validation_failed = false;
     bool *should_be_freed = NULL;
@@ -2869,7 +2870,7 @@ dm_validate_data_info(dm_ctx_t *dm_ctx, dm_session_t *session, dm_data_info_t *i
             rc = dm_load_dependant_data(dm_ctx, session, info);
             CHECK_RC_LOG_GOTO(rc, cleanup, "Loading dependant modules failed for %s", info->schema->module_name);
 
-            if (0 != lyd_validate(&info->node, LYD_OPT_STRICT | LYD_OPT_WHENAUTODEL | LYD_OPT_CONFIG, info->schema->ly_ctx)) {
+            if (0 != lyd_validate_modules(&info->node, &info->schema->module, 1, LYD_OPT_STRICT | LYD_OPT_WHENAUTODEL | LYD_OPT_CONFIG)) {
                 SR_LOG_DBG("Validation failed for %s module", info->schema->module->name);
                 validation_failed = true;
             } else {
@@ -2906,6 +2907,13 @@ dm_validate_data_info(dm_ctx_t *dm_ctx, dm_session_t *session, dm_data_info_t *i
             rc = dm_get_tmp_ly_ctx(dm_ctx, info->required_modules, &tmp_ctx);
             CHECK_RC_MSG_GOTO(rc, cleanup, "Failed to acquire tmp ctx");
 
+            /* get module from tmp_ctx */
+            mod = ly_ctx_get_module(tmp_ctx->ctx, info->schema->module->name, NULL, 1);
+            if (!mod) {
+                SR_LOG_ERR("Failed to find module '%s' in temtporary context", info->schema->module->name);
+                goto cleanup;
+            }
+
             /* migrate data to working context */
             for (size_t i = 0; i < data_for_validation->count; i++) {
                 dm_data_info_t *d = (dm_data_info_t *) data_for_validation->data[i];
@@ -2923,7 +2931,7 @@ dm_validate_data_info(dm_ctx_t *dm_ctx, dm_session_t *session, dm_data_info_t *i
             }
 
             /* start validation */
-            if (0 != lyd_validate(&data_tree, LYD_OPT_STRICT | LYD_OPT_WHENAUTODEL | LYD_OPT_CONFIG, tmp_ctx->ctx)) {
+            if (0 != lyd_validate_modules(&data_tree, &mod, 1, LYD_OPT_STRICT | LYD_OPT_WHENAUTODEL | LYD_OPT_CONFIG)) {
                 SR_LOG_DBG("Validation failed for %s module", info->schema->module->name);
                 validation_failed = true;
             } else {
@@ -2939,7 +2947,7 @@ dm_validate_data_info(dm_ctx_t *dm_ctx, dm_session_t *session, dm_data_info_t *i
             info->node = sr_dup_datatree_to_ctx(data_tree, info->schema->ly_ctx);
         }
     } else {
-        if (0 != lyd_validate(&info->node, LYD_OPT_STRICT | LYD_OPT_WHENAUTODEL | LYD_OPT_CONFIG, info->schema->ly_ctx)) {
+        if (0 != lyd_validate_modules(&info->node, &info->schema->module, 1, LYD_OPT_STRICT | LYD_OPT_WHENAUTODEL | LYD_OPT_CONFIG)) {
             SR_LOG_DBG("Validation failed for %s module", info->schema->module->name);
             validation_failed = true;
         } else {
@@ -3475,6 +3483,8 @@ dm_validate_module_inv_data_deps(dm_ctx_t *dm_ctx, dm_session_t *session, sr_err
     sr_llist_node_t *ll_node = NULL;
     md_dep_t *dep = NULL;
     bool validation_failed = false;
+    bool is_enabled = false;
+    sr_datastore_t ds = session->datastore;
 
     md_ctx_lock(dm_ctx->md_ctx, false);
     rc = md_get_module_info(dm_ctx->md_ctx, module_name, NULL, NULL, &module);
@@ -3488,8 +3498,18 @@ dm_validate_module_inv_data_deps(dm_ctx_t *dm_ctx, dm_session_t *session, sr_err
         if (dep->type != MD_DEP_DATA || !dep->dest->has_data) {
             continue;
         }
+
+        rc = dm_has_enabled_subtree(dm_ctx, dep->dest->name, NULL, &is_enabled);
+        CHECK_RC_LOG_GOTO(rc, cleanup, "Failed to check whether module is enabled for %s", dep->dest->name);
+        if (!is_enabled && (ds == SR_DS_RUNNING)) {
+            /* use startup data in this case */
+            session->datastore = SR_DS_STARTUP;
+        }
+
         rc = dm_get_data_info_internal(dm_ctx, session, dep->dest->name, true, &must_free_info, &info);
         CHECK_RC_LOG_GOTO(rc, cleanup, "Failed to load data info for %s", dep->dest->name);
+
+        session->datastore = ds;
 
         /* The dependent data has changed, so leaf refs might not be valid anymore */
         dm_invalidate_leaf_refs(info->node);

